### ISSUE TITLE
AssertionError

### ISSUE THREAD

**Issue by LancerXE** [rayguan97/GANav-offroad#16]
Hi, thanks for this great work!
I met the AssertionError when I was trying to test my model:

```
2022-07-28 10:42:54,233 - mmseg - INFO - OpenCV num_threads is `<built-in function getNumThreads>
2022-07-28 10:42:54,233 - mmseg - INFO - Loaded 791 images
load checkpoint from local path: ./trained_models/unreal/iter_20000.pth
./tools/test.py:252: UserWarning: SyncBN is only supported with DDP. To be compatible with DP, we convert SyncBN to BN. Please use dist_train.sh which can avoid this error.
  'SyncBN is only supported with DDP. To be compatible with DP, '
[                                                  ] 0/791, elapsed: 0s, ETA:Traceback (most recent call last):
  File "./tools/test.py", line 306, in <module>
    main()
  File "./tools/test.py", line 269, in main
    format_args=eval_kwargs)
  File "c:\ganav-offroad-new\mmseg\apis\test.py", line 91, in single_gpu_test
    result = model(return_loss=False, **data)
  File "C:\Anaconda\envs\ganav\lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "C:\Anaconda\envs\ganav\lib\site-packages\mmcv\parallel\data_parallel.py", line 50, in forward
    return super().forward(*inputs, **kwargs)
  File "C:\Anaconda\envs\ganav\lib\site-packages\torch\nn\parallel\data_parallel.py", line 166, in forward
    return self.module(*inputs[0], **kwargs[0])
  File "C:\Anaconda\envs\ganav\lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "C:\Anaconda\envs\ganav\lib\site-packages\mmcv\runner\fp16_utils.py", line 110, in new_func
    return old_func(*args, **kwargs)
  File "c:\ganav-offroad-new\mmseg\models\segmentors\base.py", line 110, in forward
    return self.forward_test(img, img_metas, **kwargs)
  File "c:\ganav-offroad-new\mmseg\models\segmentors\base.py", line 92, in forward_test
    return self.simple_test(imgs[0], img_metas[0], **kwargs)
  File "c:\ganav-offroad-new\mmseg\models\segmentors\encoder_decoder.py", line 264, in simple_test
    seg_logit = self.inference(img, img_meta, rescale)
  File "c:\ganav-offroad-new\mmseg\models\segmentors\encoder_decoder.py", line 247, in inference
    seg_logit = self.whole_inference(img, img_meta, rescale)
  File "c:\ganav-offroad-new\mmseg\models\segmentors\encoder_decoder.py", line 209, in whole_inference
    seg_logit = self.encode_decode(img, img_meta)
  File "c:\ganav-offroad-new\mmseg\models\segmentors\encoder_decoder.py", line 75, in encode_decode
    out = self._decode_head_forward_test(x, img_metas)
  File "c:\ganav-offroad-new\mmseg\models\segmentors\encoder_decoder.py", line 101, in _decode_head_forward_test
    seg_logits= self.decode_head.forward_test(x, img_metas, self.test_cfg)
  File "c:\ganav-offroad-new\mmseg\models\decode_heads\ours_head_class_attn.py", line 194, in forward_test
    out, maps = self.forward(inputs)
  File "c:\ganav-offroad-new\mmseg\models\decode_heads\ours_head_class_attn.py", line 138, in forward
    out, attn = self.attn(x)
  File "C:\Anaconda\envs\ganav\lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "c:\ganav-offroad-new\mmseg\models\backbones\transnet.py", line 179, in forward
    assert h==self.h
AssertionError
```
I'm using the new code updated on 25th of July.
I think the error is the same as a previous issue #2, so I added this line of code in `test=dict()` in the config file:
`dict(type='Pad', size=(300, 375), pad_val=0, seg_pad_val=255),`
Then the error disappered, but I found that the result iamges were vertically compressed and has a black border at the bottom (compared with the original image):
![捕获](https://user-images.githubusercontent.com/98766249/181488506-8a2b5d57-24e9-4f3f-b1e6-18f2caca2368.PNG)
I'm not sure if it's because the padding, cause when I tried to run the 'visulize.py' in the previous code and there is no black border of the tested image:
![pred](https://user-images.githubusercontent.com/98766249/181489226-99a8ac71-dd9d-4280-8f58-a0aa7586fabc.png)

**Comment by LancerXE** [comment]
I also met this two errors when I tried to test the RUGD dataset with the pre-trained model:
`KeyError: 'RUGDDataset_Group is not in the dataset registry'`
So I changed the code of the config file 'ganav_rugd_6' at line 71, 116, 138 and line from `RUGDDataset_Group`  to `RUGDDataset_Group6`, and it works.
Another error is :
`KeyError: "EncoderDecoder: 'OursHeadClassAttNew is not in the models registry'"`
For this, I changed the code in the same file at line 20 from `OursHeadClassAttNew` to `OursHeadClassAtt`. Then the testing code is ready to run.
Hope it helps

**Comment by rayguan97** [comment]
Thanks for the catch! The config is only updated in the trained_model folder but not updated on the google drive. This has been fixed. 

Regarding your Q1, I'm actually not sure why this happened. What is your current resolution on the test image? I think you might have to choose a good padding for your own dataset (not too disproportional), if you are training from scratch. For example, if you want to use trained RUGD model to test on your own image, you have to use the same padding size for inference ((300, 375) for RUGD). But you have training image and labels, you can change the padding in the training and testing pipeline that fit you data best.

**Comment by LancerXE** [comment]
Hi, Thanks for the quick reply!
My current resolution on the test image is 960*540. It's a small dataset trained it based on the pre-trained model of the RUGD dataset, I didn't change the padding during the training process (it's still (300,375)), should I change it in this case?

**Comment by rayguan97** [comment]
I don’t have a good answer to this since I have not used this size. You can try both and see which works for you best. If you have a small dataset, I assume it would not generalize well.

I’m closing this issue since it’s regarding assertion error. Please open another issue if you have more questions.

**Comment by LancerXE** [comment]
Yeah, I understand.
I tried to change the crop size in the code, but I always get the AssertionError when the crop size is not (300, 375), I'm not sure if it's because the 'RUGDDataset_Group6' dataset type only compatiable with (300, 375). 
And May I ask how did you define the dataset type (line 2) in **/base/dataset/rugd_group6.py**, it seems that the mmcv will check if this string is in the library. I'm wondering if the type of dataset is related to the crop size.

**Comment by rayguan97** [comment]
You can search for “ RUGDDataset_Group6” globally to figure out how to define the dataset. You need to create your own py files in base and config folder. 

The provided model weight only takes 300 by 375 patches. But if you train your own model, the model weight can be anything. Everything including the dataset is define by yourself, so you need to walk through the code, figure out the details and see how to define your own dataset. You should also check the assertion and see what it’s checking. Another good source is from the original mmseg GitHub(https://github.com/open-mmlab/mmsegmentation/blob/master/docs/en/tutorials/customize_datasets.md). The structure of our code is based on mmseg, but only support our method.

**Comment by rayguan97** [comment]
The link provided just to help you understand the structure of the config definition, but we also made some changes in our repo as well. As to the assertion, as long as the training patch size and testing patch size are the same, there should be no error. (If you train with crop size (x, y), the model is created with weights that fits (x,y) patches, so there would be no problem during training and testing. I have not personally test other size, but I started from rugd and generalize the crop size to rellis-3D, so there should be no problem.
